### PR TITLE
Added the ability to automatically use protos in nuget packages, and to easily package protos with nuget

### PR DESCRIPTION
**Pull request by stueeey**
See https://github.com/grpc/grpc-dotnet/issues/183 

Protobuf items can now be marked for packing as server, client or both, and any project using the resulting package will automatically generate partial implementations

To package a proto, set the GRPCPackAs prop on your protobuf

**This can also be done via the UI in VS**
image

To pack the project with only protos (no other files): 
     `dotnet pack /p:PackOnlyProtos="true"`
Or with a different package id (if the project already emits a package):
      `dotnet pack /p:PackOnlyProtos="true" /p:PackageId="YourPackage.Protos"`

Generating server/client from a nuget package is as simple as referencing it and `Grpc.Tools`

**Sample build output**
image

@AspirinSJL

### CODE REVIEW

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @stueeey isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Comment by stueeey** [comment]
Have signed the CLA, how do I add tags?

**Comment by stueeey** [comment]
@jtattermusch do I need to do anything further here before it can be reviewed?

**Comment by jtattermusch** [comment]
@stueeey only maintainers can add tags, I've done so.
I will try to review soon.

**Comment by jtattermusch** [comment]
CC @JunTaoLuo  @JamesNK

**Comment by JamesNK** [comment]
@jtattermusch I've seen a lot of discussion about this but I haven't had time to look into it closely, and I'm not an expert when it comes to build files. Afraid I can't contribute much to this one!

**Comment by stueeey** [comment]
@jtattermusch Please sir? 

Also the CI system seems very upset since my merge commit, did I do something wrong?

**Comment by jtattermusch** [comment]
@stueeey  for external contribution, we need to whitelist the PR so tests can run (I've done so a minute ago).

File: src/csharp/Grpc.Tools/build/_grpc/_Grpc.Tools.targets [hunk 1]
```diff
@@ -29,6 +31,101 @@
...
+    <Message Text="Found @(_ServerProtoNugetFilesFound-&gt;Count()) packaged server proto(s):" Importance="Normal" Condition="'$(_FoundServerProtoFiles)' == 'True'" />
+    <Message Text="	%(_ServerProtoNugetFilesFound.Identity)" Importance="Normal" Condition="'$(_FoundServerProtoFiles)' == 'True'" />
+    <Message Text="Found @(_BothProtoNugetFilesFound-&gt;Count()) packaged client &amp; server proto(s):" Importance="Normal" Condition="'$(_FoundBothProtoFiles)' == 'True'" />
+    <Message Text="	%(_BothProtoNugetFilesFound.Identity)" Importance="Normal" Condition="'$(_FoundBothProtoFiles)' == 'True'" />
+  </Target>
+
+  <!-- 
+    This target will mark any Protobuf with GrpcPackAs set to Server, Client or Both as packable
+    and will set the path in the nuget package to one of the following:
+    protobuf/grpc
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/build/_grpc/_Grpc.Tools.targets]: using this convention seems very arbitrary. What makes you think that the uses will want to adopt this layout for their projects?

**Comment by jtattermusch** [comment]
Thanks for the contributions. TBH don't understand the details of how this is supposed to be used.
Can you explain the use case a bit more. Also, because this seems to be a feature proposal where people will have different opinions, it would be better to start with explaining the use case and the approach used first, before actually implementing this (the MSBuild diffs aren't very readable and it's hard to understand how this should be used).

**Comment by jtattermusch** [comment]
My guess what the intended use case could be (please confirm or explain)
- the user implements a library that contains .proto files (besides .cs files) and wants to create a nuget out of it
- when packing the project, a nuget is created that only contains the .proto files (and no code-generated code). It also contains metadata about what the intended protoc options are going to be.
- in another project, someone pulls the nuget (with .proto files) as a dependency. If they also add Google.Protobuf.Tools package (not added by default), the protobuf C# code will be generated for them automatically, using the options provided in the nuget.

If this description is accurate, I think there will be some questions (it's not as simple as it looks like and there's some corner cases IMHO).  I'd also like to understand why this approach is better than simply including the generated protobuf C# code in the nuget package.

File: src/csharp/Grpc.Tools/build/_grpc/_Grpc.Tools.targets (hunk 1 above)
Comment (stueeey) [reply src/csharp/Grpc.Tools/build/_grpc/_Grpc.Tools.targets]: The convention is for the layout of the nuget package, not the project itself (see screenshot in pr description)

**Comment by stueeey** [comment]
Your summary of the intention is pretty close.

The primary intention is to make packaging proto files as easy as possible and preferably not requiring the dark art of msbuild as it is not a common skill (see linked issue)

This change would allow you to build a server app with a proto in it, mark it as publish and then just pack the project to get a client, server or client + server package. Then the package can just be pulled in by the client and they are good to go. Because the code is being compiled into the client it can also use the advantages of having a partial class and extend it. Do you think we need a sample project to Illustrate this?

The package actually doesn't specify the protoc options beyond whether to generate a client or server, the consuming client project can specify the rest (see Protobuf_DefaultAccess and Protobuf_DefaultClientBaseType)

However you did point out something I missed which is that the emitted package should probably have a dependancy on GRPC.tools. what are the corner cases you are concerned about?

**Comment by Chelaris182** [comment]
could you please share the status of the PR?

**Comment by stueeey** [comment]
Stalled..

**Comment by simonpinn** [comment]
Hey @stueeey - have just run into the same issue as faced in https://github.com/grpc/grpc-dotnet/issues/183 which is how I find myself here.... any progress?

**Comment by jtattermusch** [comment]
While this is a useful addition in principle, we'd need to improve our testing story for Grpc.Tools before we can accept improvements like this (the existing Grpc.Tools logic is already quite complex and it's very fragile because the combination of using non-trivial msbuild scripts and the protoc/codegen plugins being native components makes it very challenging to come up quick, reliable and easy-to-maintain tests).

So for now, I'd like to put this PR on hold. We can revive in the future if we have more solid foundations to build on top of.

**Pull request closed without merging**